### DISCUSSION TITLE
Help writing to a log file

### DISCUSSION THREAD

**Discussion by sidprice** [rokath/trice#333] [Q&A] [answered]
First, thank you for your great effort on this project. It looks most useful.

I have a device that has and internal file system and I would like to output the Trice messages to there in files. The device also has an RTOS, so my plan is to have a task specifically for writing Trice messages to the file system. It is not clear to me which functions I need to rewrite in order to send the Trice messages to a file, could you advise please?

**Comment by rokath** [comment]
Hi sidprice, what I understand from your question: It is only one single device. You intend to use TRICE  macros on that device and wish to have a log file on the very same device on the device internal file system.
The question is what you intend to do with this internal logfile. I assume you want be able to see the logs later on a PC. If this is your Intention, I have a proposal for you, coming soon.

**Comment by sidprice** [comment]
@rokath Yes, the device logs will be kept locally and then may be transferred via Bluetooth or by the PC reading the filesystem via USB.

**Comment by rokath** [comment, marked as answer]
Inside the `triceConfig around line 48 you see this:
```C
//#define TRICE_RTT_CHANNEL 0 //!< Enable and set channel number for SeggerRTT usage. Only channel 0 works right now for some reason.
#define TRICE_UART USART2 //!< Enable and set UART for serial output.
```
Proposal: Change that to:
```C
//#define TRICE_RTT_CHANNEL 0 //!< Enable and set channel number for SeggerRTT usage. Only channel 0 works right now for some reason.
//#define TRICE_UART USART2 //!< Enable and set UART for serial output.
#define TRICE_LOCAL_BINARY_LOGFILE //!< Enable and set a local file as logfile for output.
```
```C
#if defined( TRICE_LOCAL_BINARY_LOGFILE ) && !defined( TRICE_HALF_BUFFER_SIZE ) // direct out to local files system
#define TRICE_WRITE( buf, len ) do{ WriteToLocalLogFile( buf, len ); }while(0)
void WriteToLocalLogFile( uint8_t const * buf, unsigned len );
#endif
```
Alternatively:
```C
#if defined( TRICE_LOCAL_BINARY_LOGFILE ) && defined( TRICE_HALF_BUFFER_SIZE ) // buffered out to local files system
#define TRICE_WRITE( buf, len ) do{ WriteToLocalLogFile( buf, len ); }while(0)
void WriteToLocalLogFile( uint8_t const * buf, unsigned len );
#endif
```

Then provide function `WriteToLocalLogFile` in your code, which writes to a log file, after you opened it.

After you transferred `trices.raw` or `MyTrices.bin` to the PC, you can do:
```bash
trice log -port FILEBUFFER 
```
assuming `til.json` and `trices.raw` in the current directory or:
```bash
trice log -port FILEBUFFER -args "path/to/MyTrices.bin" -idList "path/to/til.json"
```
If you mount the target filesystem  to the PC you can also say:
```bash
trice log -port FILE -args "mnt/path/to/trices.raw" -idList "mnt/path/to/til.json"
```
to see the logs continously.

This proposal logs the trice messages in binary form. A log in text form I would not recommend for space and speed reasons.
It could make sense to store `til.json` (and maybe `li.json`) on the target file system.
